articles/how-to-setup-user-authentication-from-scratch-with-rails-6/index.md

```diff
@@ -0,0 +1,702 @@
+
+- touch `app/controllers/sessions_controller.rb`
+
+```rb
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:email])
+    # finds existing user, checks to see if user can be authenticated
+    if user.present? && user.authenticate(params[:password])
+    # sets up user.id sessions
+      session[:user_id] = user.id
+      redirect_to root_path, notice: 'Logged in successfully'
+    else
+      flash.now[:alert] = 'Invalid email or password'
+      render :new
+    end
+  end
+
+  def destroy
+    # deletes user session
+    session[:user_id] = nil
+    redirect_to root_path, notice: 'Logged Out'
+  end
+end
+```
+`SessionsController` provides login functionality to an existing user, also logs out a user by deleting session data.
+
+- touch `app/controllers/passwords_controller.rb`
+
+```rb
+class PasswordsController < ApplicationController
+  # allows only logged in users
+  before_action :require_user_logged_in!
+
+  def edit; end
+
+  def update
+    # update user password
+    if Current.user.update(password_params)
+      redirect_to root_path, notice: 'Password Updated'
+    else
+      render :edit
+    end
+  end
+
+  private
+  def password_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+end
+```
+`PassWordsController` allows signed_in_user to update passwords.
+
+Update `app/controllers/application_controller.rb`
+
+```rb
+class ApplicationController < ActionController::Base
+  before_action :set_current_user
+
+  def set_current_user
+    # finds user with session data and stores it if present
+    Current.user = User.find_by(id: session[:user_id]) if session[:user_id]
+  end
+
+  def require_user_logged_in!
+    # allows only logged in user
+    redirect_to sign_in_path, alert: 'You must be signed in' if Current.user.nil?
+  end
+end
+```
+- Note inheritance hierarchy in this class, `set_current_user` will be accessed with all controllers in the subdirectory.
+  This controller finds signed_in_user with `session[:user_id]` and stores it as `Current.user` if present and can be accessed in our views.
+
+- touch `app/models/current.rb`
+
+```rb
+class Current < ActiveSupport::CurrentAttributes
+  # makes Current.user accessible in view files.
+  attribute :user
+end
+```
+
+### Configuring views
+
+Controllers make model data available to the view,this data can be displayed to the user.
+
+Create a `sign_up` form
+- touch `app/views/registrations/new.html.erb`
+
+```erb
+<h1>Sign Up</h1>
+
+<%= form_with model: @user, url: sign_up_path do |f| %>
+  <p>
+  <%= f.label 'email' %><br>
+  <%= f.text_field :email %>
+  </p>
+
+  <p>
+  <%= f.label 'password' %><br>
+  <%= f.password_field :password %>
+  </p>
+
+  <p>
+  <%= f.label 'password_confirmation' %><br>
+  <%= f.password_field :password_confirmation %>
+  </p>
+
+  <p>
+  <%= f.submit 'Sign Up' %>
+  </p>
+<% end %>
```

Comment:
Explain what happens here. You can just explain about forms once here and ignore the other forms.